modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+= Type System
+ifndef::env-site,env-github[]
+include::_attributes.adoc[]
+endif::[]
+
+DataWeave 2.0 introduced a new type system on the language. To understand how values for these types are defined see xref:dataweave-types.adoc[Value Constructs for DataWeave Types] documentation.
+
+A type system defines a set of constraints to a set of constructs such as:
+
+* Variables
+* Function parameters
+
+These constraints are used in the type checking phase where DataWeave ensures that the values assigned
+to variables or the arguments for a function call respects its constraints.
+
+For example:
+
+----
+var userName: String = "John"
+----
+
+The variable definition for `userName` has the constraint of assigning a value of type String.
+If some other type value is assigned, a type checking error raises.
+
+----
+fun toUser(id: Number, userName: String): String = "you called the function toUser!"
+----
+
+The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types) and the result of the function has to be of type String.
+
+----
+toUser(123, "John")
+toUser("123", true)
+----
+
+Both calls don't throw an exception. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +
```

Comment:
@afelisatti Would the above sentence replace the entire paragraph then? Meaning that this is not necessary: _the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"`_

Reply:
No, this would be more like an addition. Something like `Though you would expect the second call to fail, autocoercions allow it to be successful and both of them work. In the case of the second function call, there is an autocoercion of the String ...`

Reply:
Updated